README.md
# GaMorNet (Galaxy Morphology Network)

GaMorNet is a Convolutional Neural Network based on AlexNet to classify galaxies morphologically. GaMorNet does not need a large amount of training data (as it is trained on simulations and then transfer-learned on a small portion of real data) and can be applied on multiple datasets. Till now, GaMorNet has been tested on ~100,000 SDSS g-band galaxies and ~20,000 CANDELS H-band galaxies and has a misclassification rate of <5%. 

The training, testing and relevant statistics of GaMorNet is outlined in the paper mentioned in the next section. Please refer to the relevant sections of the paper for details about the training and transfer learning methods. 

## Usage Info/Citation/Reference
This repository was used in the work pertaining to the following research article:-
"Galaxy Morphology Network (GaMorNet):  A Convolutional Neural Network used to study morphology and quenching in ∼100,000 SDSS and ∼20,000 CANDELS galaxies", Ghosh et. al.

If you use this code for any published work, please cite the above paper and include a link to this GitHub repository.

This code is being made available under a GNU General Public License v3.0. Please see the file called LICENSE in the repository for more details.

---

## What do you need to run GaMorNet?
GaMorNet was coded in Python using [TFLearn](http://tflearn.org/) which is a high-level API for [Tensorflow](https://www.tensorflow.org/). GaMorNet has been tested to work with the following versions of different libraries mentioned below.

| Python  |  Numpy | Tensorflow-gpu  |  TFLearn  | CUDA  | cuDNN | 
|---|---|---|---|---| --- |
| 2.7.14 | 1.16.14 | 1.12.0 | 0.3.2 | 9.0.176 | 7.1.4 |
| 3.6.9  | 1.17.0  | 1.12.0  | 0.3.2  | 9.0.176 | 7.1.4 |
| 3.6.9 | 1.17.0 | 1.13.1 | 0.3.2 | 10.0.130 | 7.6.0 |

**The first configuration is what GaMorNet was originally developed in and thus is the most stable; however substantial testing has also been performed with the other two configurations**. It is known that the last configuration mentioned leads to some Tensorflow depreciation warnings.

Steps to install GaMorNet dependencies:-

* It is highly recommended to initiate a Python virtual environment (eg. using [Anaconda](https://www.anaconda.com/distribution/)) with the above-mentioned versions of Python, Numpy, TF-gpu, and TFLearn. Note that CUDA and cuDNN are necessary if you want to use GPU acceleration. More information on using Tensorflow GPU acceleration is available [here](https://www.tensorflow.org/install/gpu). Some other Python libraries that GaMorNet depends on are matplotlib, astropy, math, time and multiprocessing.

   * To initiate a new conda environment with a specific version of Python, you can do `conda create -n yourenvname python=x.x` where x.x is the Python version number and yourenvname is the name of the environment
   * Then, activate the environment using `conda activate yourenvname` or ` source activate yourenvname`
   * To install the relevant numpy version, do `conda install numpy=x.x.x` where x.x.x is the appropriate version number
   * For installing the other libraries, the following commands should suffice:-
   ```
   conda install matplotlib
   pip install astropy
   pip install multiprocessing
   ```

   * Now, install Tensorflow in the same environment. [Instructions for Installing Tensorflow](https://www.tensorflow.org/install) [**Please don't install Tensorflow in a separate environment**] 

   * Now, install TFLearn in the same environment. [Instructions for Installing TFLearn](http://tflearn.org/installation/) (Recommended way is to just do `pip install tflearn`) [**Please don't install TFLearn in a separate environment**]

   * To make sure that you have installed both TFLearn and Tensorflow correctly, run the following piece of code in an interactive Python session to verify the installation. 

   ```
   python
   import tensorflow as tf
   hello = tf.constant('Hello, TensorFlow!')
   sess = tf.Session()
   print(sess.run(hello))
   a = tf.constant(10)
   b = tf.constant(32)
   print(sess.run(a + b))

   import tflearn as tfl
   ```
   * Finally to de-activate your environment, type `conda deactivate` or `source deactivate`

* If all the above commands work, then you are all set and the gamornet scripts should run without any errors. **If there are  warnings or errors, please check to make sure that you have the recommended versions of critical libraries according to the table above**

---

## The Network

### Design -- [gamornet.py](gamornet.py)
The file [gamornet.py](gamornet.py) contains the code that we use to create the GaMorNet model in TFLearn. *Note that this file will not run without further modification -- this is only meant to show what exact TFLearn functions we used to code GaMorNet*

### Using our Trained Models --- [gamornet_predict.py](/gamornet_predict/gamornet_predict.py)
In our paper, we outline how we train GaMorNet. These trained models can be accessed via http://www.astro.yale.edu/aghosh/gamornet.html or http://gamornet.ghosharitra.com . In order to use these trained models, we provide some example code in the file [gamornet_predict.py](/gamornet_predict/gamornet_predict.py)

For this demonstration, we will be using our final trained SDSS model to predict the classification of two randomly chosen SDSS g-band images stored in the directory [sdss_cutouts](/gamornet_predict/sdss_cutouts/). Positional information on these galaxies is available in the [info.txt](/gamornet_predict/sdss_cutouts/info.txt) file.

The following steps demonstrate what you need to do, to run the script successfully. 

* Clone this GitHub Repository using the appropriate  https or ssh link
    * ```git clone https://github.com/aritraghsh09/GaMorNet.git``` 
    * OR
    * ``` git clone git@github.com:aritraghsh09/GaMorNet.git```
    
* Make sure that you have the following Python Libraries already installed besides Tensorflow and TFLearn :- numpy, pylab, astropy, math, time, multiprocessing

* Download the 3 model files using this ftp link ftp://ftp.astro.yale.edu/pub/aghosh/gamornet/trained_models/SDSS/tl/ and **store these in the [gamornet_predict](/gamornet_predict/) directory**
   * You can copy & paste the above ftp link into your browser and them manually download and place the relevant files in the correct directory.
   * OR use the following commands using a terminal
   * `cd GaMorNet/gamornet_predict/`
   * `ftp ftp.astro.yale.edu`
   * If you are prompted for a username, enter ```anonymous``` and keep the password field blank
   * After logging in, navigate to the appropriate directory using ```cd pub/aghosh/gamornet/trained_models/SDSS/tl/```
   * `get check-1546293.meta`
   * `get check-1546293.index`
   * `get check-1546293.data-00000-of-00001`
   * `quit`

* Activate the anaconda environment (where you installed the GaMorNet dependencies) using `conda activate yourenvname` or ` source activate yourenvname`

* Run the [gamornet_predict](/gamornet_predict/gamornet_predict.py) script using ```python gamornet_predict.py```

* The predicted probabilities for the test images should be written to an output file named predictions.txt


To run predictions on other images than the ones supplied, keep the following information in mind:-

* GaMorNet-SDSS & GaMorNet-CANDELS were trained for square images of 167 pixels and 83 pixels respectively. If you are using our trained models, you need to make sure the cutouts you are using are of the same size. The code will still run if you use cutouts of a different size (as the input data is reshaped to the appropriate size) but will give you erroneous results. 

* GaMorNet-SDSS & GaMorNet-CANDELS were trained for galaxies at z\~0 and z\~1 respectively. If you are using our trained models, you need to make sure the cutouts you are using are at similar redshifts. To perform predictions on galaxies at a substantially different z, you will need to retrain GaMorNet accordingly. 

* GaMorNet-SDSS & GaMorNet-CANDELS were trained with g-band and H-band images respectively. If you are using our trained models, you need to make sure the galaxy images you are using are in nearby bands. To perform predictions on galaxies at a substantially blueshifted or redshifted band, you might need to retrain GaMorNet accordingly. 

* To perform prediction on CANDELS images, you need to alter the following parameters in the code
  * In the ```array_image``` function in [gamornet_predict.py](/gamornet_predict/gamornet_predict.py) alter the ```newshape``` argument to ```newshape=(83,83,1)```
  * The input layer of the network needs to be changed to ```network = input_data(shape=[None, 83, 83, 1])```
  * You need to download the appropriate CANDELS trained models instead of the SDSS models mentioned above  
  
### (Re)Training GaMorNet -- [gamornet_train.py](/gamornet_train/gamornet_train.py)
In our paper, we outline how we first train GaMorNet on simulated images and then transfer-learn (i.e. retrain the network) on real images. 

The script, [gamornet_train.py](/gamornet_train/gamornet_train.py) is meant to demonstrate how GaMorNet can be trained. For demonstrating this, we will be using 5 simulated SDSS images from our sample. These images are stored in the folder [simulated_images](/gamornet_train/simulated_images/). The parameters used to simulate these galaxies is stored in the file [sim_para.txt](/gamornet_train/simulated_images/sim_para.txt). Using this file, we deduce the correct classification for each galaxy by using the integrated magnitudes of the disk and bulge for each galaxy. 

*Note that the script is set up to train for 5 epochs; use 3 images for training and 2 for validation. This is for demonstration purposes and will not lead to any useful results. You need to use much larger number of simulated images (~100,000) and train for many more epochs.(~500)*

The following steps demonstrate what you need to do, to run the script successfully. 

* Clone this GitHub Repository using the appropriate  https or ssh link
    * ```git clone https://github.com/aritraghsh09/GaMorNet.git``` 
    * OR
    * ``` git clone git@github.com:aritraghsh09/GaMorNet.git```
  
* Activate the anaconda environment (where you installed the GaMorNet dependencies) using `conda activate yourenvname` or ` source activate yourenvname`

* Make sure that you have the following Python Libraries already installed besides Tensorflow and TFLearn :- numpy, pylab, astropy, math, time, multiprocessing

* Run the script [gamornet_train.py](/gamornet_train/gamornet_train.py) using `python gamornet_train.py`. 

* The model file will be stored as 3 files named checkpoint-5.meta, checkpoint-5.index, checkpoint-5.data-00000-of-00001 in the /gamornet_train/ directory. The number 5 indicates that the model was trained for 5 epochs.


#### To retrain starting from one of our models
In this demo., we will be retraining GaMorNet using the above mentioned simulated images starting from our final SDSS model. *Retraining our final model on simulated images doesn't make any scientific sense and is just for demonstration purposes.* 

* Download the 3 model files using this ftp link ftp://ftp.astro.yale.edu/pub/aghosh/gamornet/trained_models/SDSS/tl/ and **store these in the [gamornet_train](/gamornet_train/) directory**
   * You can copy & paste the above ftp link into your browser and then manually download and place the relevant files in the correct directory.
   * OR use the following commands using a terminal
   * `cd GaMorNet/gamornet_train/`
   * `ftp ftp.astro.yale.edu`
   * If you are prompted for a username, enter ```anonymous``` and keep the password field blank
   * After logging in, navigate to the appropirate directory using ```cd pub/aghosh/gamornet/trained_models/SDSS/tl/```
   * `get check-1546293.meta`
   * `get check-1546293.index`
   * `get check-1546293.data-00000-of-00001`
   * `quit`
   
* Activate the anaconda environment (where you installed the GaMorNet dependencies) using `conda activate yourenvname` or ` source activate yourenvname`

* Uncomment Line #138 in the GaMorNet code ``` #model.load(modelLoadPath+"check-1546293")```

* If you want to freeze/lock some layers during training, add the `trainable=False` argument to the layer you want to lock. For eg. to lock Layer 8, uncomment the following line `#network = conv_2d(network, 384, 3, activation='relu',trainable=False)  ` and **comment out the original layer 8 (Line #105)**

* If you don't want load some layers from the model and want to initialize these from scratch, add the `restore=False` argument to the layer you don't want to load. For eg. to not load Layer 13, uncomment the following line ` #network = fully_connected(network, 4096, activation='tanh',restore=False)` and **comment out the original layer 13 (Line #115)**

* Run the script [gamornet_train.py](/gamornet_train/gamornet_train.py) using `python gamornet_train.py`. 

* The model file will be stored as 3 files named checkpoint-1546298.meta, checkpoint-1546298.index, checkpoint-1546298.data-00000-of-00001 in the /gamornet_train/ directory. The number 1546298 indicates that the model was trained for 5 epochs since 1546293

---
## Where are all the Model Files? 
The Base Directory for all the models is [ftp://ftp.astro.yale.edu/pub/aghosh/gamornet/trained_models/](ftp://ftp.astro.yale.edu/pub/aghosh/gamornet/trained_models/)

After that the different models are arranged as follows:-
* GaMorNet-S model trained only on simulations &rightarrow; /SDSS/sim_trained/
* GaMorNet-S model trained on simulations and then transfer learned on real data &rightarrow; /SDSS/tl/
* GaMorNet-C model trained only on simulations &rightarrow; /CANDELS/sim_trained/
* GaMorNet-C model trained on simulations and then transfer learned on real data &rightarrow; /CANDELS/tl/


**For other products of the public data release, please refer to the Appendix of the paper or head to [this link](http://www.astro.yale.edu/aghosh/gamornet.html) or [this link](http://gamornet.ghosharitra.com).**

--- 
## Questions?
* If you have an upgrade idea or some way to make the code more efficient, please go ahead and submit a pull request.
* If you are sure you followed the installation instructions correctly and there is something wrong with the instructions, please open up an issue on GitHub or submit a pull request with relevant changes to the read-me.
* If you are not sure about how to use something, please send me an email at aritraghsh09+gamornet at gmail.com 


changelog.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),

## [Unreleased]

### Added

### Changed

### Fixed


## [0.4.3] -- 2021-07-14

### Changed
- Changed setup.py to force h5py installation of <3.0. 
This is because of this error -- https://github.com/keras-team/keras/issues/14265

## [0.4.2] -- 2020-07-14

### Changed/Fixed
- Changed setup.py to force Keras installation of <2.4 

## [0.4.1] -- 2020-06-14

### Changed/Fixed
- Corrected errors in the Python Docstrings of both keras and tflearn modules
- Corrected errors/typos in API documentation 


## [0.4] -- 2020-06-12

### Added
- Documentation and Tutorials
- Code is now Production Ready as Well



## [0.3.1] - 2020-06-10

### Fixed
- gamornet_predict_tflearn can handle the situation now when batch_size > number of images



## [0.3] - 2020-06-04

### Added
- Basic Documentation for the Project

### Changed
- updated LRN function in keras_module to remove dependancy on image_dim_ordering function of keras backend
- altered the naming scheme of saved tflearn models.
- increased the flushing time for the gamornet_predict_tflearn function progressbar
- implemented single version and author sourcing. (both are sourced from __init__.py to setup.py

## [0.2.3] - 2020-05-19
### Added
- Initial Release of gamornet-cpu
- Version starts with 0.2.3 to match up with the GPU Version of GaMorNet


gamornet_modular/gamornet_module_keras.py
from keras.models import Sequential, load_model, clone_model
from keras.layers import Dense, Flatten, Dropout, Activation, Input
from keras.layers import Conv2D, MaxPooling2D
from keras.initializers import VarianceScaling
from keras import optimizers
from keras.callbacks import ModelCheckpoint
from keras.callbacks import CSVLogger
from keras import backend as K
from keras.engine.topology import Layer, InputSpec
import pandas as pd
import wget
import random
import string
import os
import numpy as np

############################################
##########HELPER FUNCTIONS##################

def check_input_shape_validity(input_shape):
	
	if input_shape == 'SDSS':
		input_shape = (167,167,1)
	elif input_shape == 'CANDELS':
		input_shape = (83,83,1)		  
	else:
		try:
			if len(input_shape) != 3:
				raise Exception("input_shape has to be a tuple (typically of 3 dimensions) or any of the available keywords")
		except:
			raise Exception("input_shape has to be a tuple (typically of 3 dimensions) or any of the available keywords")

	return input_shape


def check_imgs_validity(img_array):

	if isinstance(img_array,np.ndarray):
		if len(img_array.shape) != 4:
			raise Exception("The Image Array needs to have 4 dimensions. (num,x,y,bands)")
	elif isinstance(img_array,list):
		if len(np.array(img_array).shape) != 4:
			raise Exception("The Image Array needs to have 4 dimensions. (num,x,y,bands)")
	else:
		raise Exception("The Image Array Needs to be a 4 Dimensional Numpy Array.")


def check_labels_validity(labels):

	if isinstance(labels,np.ndarray):
		if labels.shape[1] != 3:
			raise Exception("The Labels Array needs to have 2 dimensions. (num,3)")
	elif isinstance(labels,list):
		if np.array(labels).shape[1] != 3:
			raise Exception("The Labels Array needs to have 2 dimensions. (num,3)")
	else:
		raise Exception("The Lables Array Needs to be a 2 Dimensional Numpy Array. (num,3)")


def check_bools_validity(bools):
	try:
		for element in bools:
			if type(element) != bool:
				raise Exception("The Supplied Array of Bools doesn't look okay")
	
		if len(bools) != 8:
				raise Exception("The Supplied Array of Bools doesn't look okay")

	except:
		raise Exception("The Supplied Array of Bools doesn't look okay")


def get_model_from_link_keras(link,model):
	
	letters = string.ascii_lowercase
	if link[-4:] == 'hdf5':
		file_name = ''.join(random.choice(letters) for i in range(15)) + '.hdf5'
	else:
		file_name = ''.join(random.choice(letters) for i in range(15)) + '.h5'
	
	wget.download(link,out=file_name)

	try:
		model.load_weights(file_name)
	except:
		os.remove(file_name)
		raise

	os.remove(file_name)
	
	return model


def gamornet_load_model_keras(model,model_load_path):
	
	if model_load_path == 'SDSS_sim':
		print("Fetching SDSS Sim Trained Weigths.....")
		link = 'ftp://ftp.astro.yale.edu/pub/aghosh/gamornet_keras/trained_models/SDSS/sim_trained/model_sdss_sim_trained.h5'
		model = get_model_from_link_keras(link,model)
		
	elif model_load_path == 'SDSS_tl':
		print("Fetching SDSS TL Weigths.....")
		link = 'ftp://ftp.astro.yale.edu/pub/aghosh/gamornet_keras/trained_models/SDSS/tl/model_sdsss_tl.hdf5'
		model = get_model_from_link_keras(link,model)
		
	elif model_load_path == 'CANDELS_sim':
		print("Fetching CANDELS Sim Trained Weigths.....")
		link = 'ftp://ftp.astro.yale.edu/pub/aghosh/gamornet_keras/trained_models/CANDELS/sim_trained/model_candels_sim_trained.hdf5'
		model = get_model_from_link_keras(link,model)
		
	elif model_load_path == 'CANDELS_tl':
		print("Fetching CANDELS TL Weigths.....")
		link = 'ftp://ftp.astro.yale.edu/pub/aghosh/gamornet_keras/trained_models/CANDELS/tl/model_candels_tl.hdf5'
		model = get_model_from_link_keras(link,model)
		
	else:
		 model.load_weights(model_load_path)
			
	return model

	
###########################################
###########################################



############################################
##########KERAS FUNCTIONS##################

#Implementing LRN in Keras. Code from "Deep Learning with Python" by F. Chollet
class LocalResponseNormalization(Layer):

	def __init__(self, n=5, alpha=0.0001, beta=0.75, k=1.0, **kwargs):
		self.n = n
		self.alpha = alpha
		self.beta = beta
		self.k = k
		super(LocalResponseNormalization, self).__init__(**kwargs)

	def build(self, input_shape):
		self.shape = input_shape
		super(LocalResponseNormalization, self).build(input_shape)

	def call(self, x, mask=None):
		if K.image_dim_ordering == "th":
			_, f, r, c = self.shape
		else:
			_, r, c, f = self.shape
		squared = K.square(x)
		pooled = K.pool2d(squared, (self.n, self.n), strides=(1, 1),
			padding="same", pool_mode="avg")
		if K.image_dim_ordering == "th":
			summed = K.sum(pooled, axis=1, keepdims=True)
			averaged = self.alpha * K.repeat_elements(summed, f, axis=1)
		else:
			summed = K.sum(pooled, axis=3, keepdims=True)
			averaged = self.alpha * K.repeat_elements(summed, f, axis=3)
		denom = K.pow(self.k + averaged, self.beta)
		return x / denom

	def get_output_shape_for(self, input_shape):
		return input_shape
	

	

def gamornet_build_model_keras(input_shape):
	
	input_shape = check_input_shape_validity(input_shape)
	
	#uniform scaling initializer
	uniform_scaling = VarianceScaling(scale=1.0, mode='fan_in', distribution='uniform', seed=None)

	#Building GaMorNet
	model = Sequential()

	model.add(Conv2D(96,11, strides=4, activation='relu', input_shape=input_shape,padding='same',kernel_initializer=uniform_scaling))
	model.add(MaxPooling2D(pool_size=3, strides=2,padding='same'))
	model.add(LocalResponseNormalization())

	model.add(Conv2D(256,5, activation='relu',padding='same',kernel_initializer=uniform_scaling))
	model.add(MaxPooling2D(pool_size=3, strides=2,padding='same'))
	model.add(LocalResponseNormalization())

	model.add(Conv2D(384, 3, activation='relu',padding='same',kernel_initializer=uniform_scaling))
	model.add(Conv2D(384, 3, activation='relu',padding='same',kernel_initializer=uniform_scaling))
	model.add(Conv2D(256, 3, activation='relu',padding='same',kernel_initializer=uniform_scaling))
	model.add(MaxPooling2D(pool_size=3, strides=2,padding='same'))
	model.add(LocalResponseNormalization())

	model.add(Flatten())
	model.add(Dense(4096, activation='tanh',kernel_initializer='TruncatedNormal'))
	model.add(Dropout(0.5))
	model.add(Dense(4096, activation='tanh',kernel_initializer='TruncatedNormal'))
	model.add(Dropout(0.5))
	model.add(Dense(3, activation='softmax',kernel_initializer='TruncatedNormal'))
	
	return model



def gamornet_predict_keras(img_array,model_load_path,input_shape,batch_size=64,individual_arrays=False):
	
	check_imgs_validity(img_array)

	model = gamornet_build_model_keras(input_shape=input_shape)
	
	print("Loading GaMorNet Model.....")
	model = gamornet_load_model_keras(model,model_load_path)
	
	print("Performing Predictions.....")
	preds = model.predict(img_array,batch_size = batch_size)
	preds = np.array(preds) #converting to a numpy array for easier handling.
	
	if individual_arrays == True:
		return preds[:,0],preds[:,1],preds[:,2] # 'disk_prob','unclass_prob','bulge_prob'
	else:
		return preds


def gamornet_train_keras(training_imgs,training_labels,validation_imgs,validation_labels,input_shape,files_save_path="./",epochs=100,checkpoint_freq=0,batch_size=64,lr=0.0001,momentum=0.9,decay=0.0,nesterov=False,loss='categorical_crossentropy',load_model=False,model_load_path="./",save_model=True,verbose=1):

	check_imgs_validity(training_imgs)
	check_imgs_validity(validation_imgs)
	check_labels_validity(training_labels)
	check_labels_validity(validation_labels)

	model = gamornet_build_model_keras(input_shape=input_shape)

	sgd = optimizers.SGD(lr=lr, momentum=momentum, decay=decay, nesterov=nesterov)
	model.compile(loss=loss, optimizer=sgd, metrics=['accuracy']) 


	callbacks_list = []
	
	if checkpoint_freq != 0:
		checkpoint = ModelCheckpoint(files_save_path + 'model_{epoch:02d}.hdf5', monitor='val_loss', verbose=verbose, save_best_only=False, save_weights_only=False, mode='auto', period=checkpoint_freq)
		callbacks_list.append(checkpoint)

	csv_logger = CSVLogger(files_save_path + "metrics.csv", separator=',', append=False)
	callbacks_list.append(csv_logger)
	
	if load_model == True:
		model = gamornet_load_model_keras(model,model_load_path)

	model.fit(training_imgs, training_labels, batch_size=batch_size, epochs=epochs, verbose=verbose, validation_data=(validation_imgs,validation_labels), shuffle=True, callbacks=callbacks_list)

	if save_model == True:
		model.save(files_save_path + "trained_model.hdf5")

	return model


	
def gamornet_tl_keras(training_imgs,training_labels,validation_imgs,validation_labels,input_shape,load_layers_bools = [True]*8,trainable_bools = [True]*8,model_load_path="./",files_save_path="./",epochs=100,checkpoint_freq=0,batch_size=64,lr=0.00001,momentum=0.9,decay=0.0,nesterov=False,loss='categorical_crossentropy',save_model=True,verbose=1):

	check_imgs_validity(training_imgs)
	check_imgs_validity(validation_imgs)
	check_labels_validity(training_labels)
	check_labels_validity(validation_labels)
	check_bools_validity(load_layers_bools)
	check_bools_validity(trainable_bools)

	model = gamornet_build_model_keras(input_shape=input_shape)
	model_new = clone_model(model)
	model = gamornet_load_model_keras(model,model_load_path)

	#Reversing the Order of the Bools because I will call .pop() on these later	
	load_layers_bools.reverse()
	trainable_bools.reverse()

	for i in range(len(model_new.layers)):
    
		if model_new.layers[i].count_params() != 0:
        
			model_new.layers[i].trainable = trainable_bools.pop()
        
			if load_layers_bools.pop() == True:
				model_new.layers[i].set_weights(model.layers[i].get_weights())
				print("Loading Layer" + str(i) + " from previous model.")
			else:
				print("Initializing Layer" + str(i)+ " from scratch")
            
		else:
			model_new.layers[i].set_weights(model.layers[i].get_weights())

	sgd = optimizers.SGD(lr=lr, momentum=momentum, decay=decay, nesterov=nesterov)
	model_new.compile(loss=loss, optimizer=sgd, metrics=['accuracy']) 

	callbacks_list = []
	
	if checkpoint_freq != 0:
		checkpoint = ModelCheckpoint(files_save_path + 'model_{epoch:02d}.hdf5', monitor='val_loss', verbose=verbose, save_best_only=False, save_weights_only=False, mode='auto', period=checkpoint_freq)
		callbacks_list.append(checkpoint)

	csv_logger = CSVLogger(files_save_path + "metrics.csv", separator=',', append=False)
	callbacks_list.append(csv_logger)
	

	model_new.fit(training_imgs, training_labels, batch_size=batch_size, epochs=epochs, verbose=verbose, validation_data=(validation_imgs,validation_labels), shuffle=True, callbacks=callbacks_list)

	if save_model == True:
		model_new.save(files_save_path + "trained_model.hdf5")

	return model_new


###########################################
###########################################
